### PR TITLE
Functional ctc loss

### PR DESCRIPTION
**Pull request by brandonliu2**
* 解决了[关于ctc_loss的几个问题](https://github.com/Oneflow-Inc/OneCloud/issues/104#issuecomment-1156122806)中的第一个`functional.ctc_loss` 未与`pytorch`对齐的问题。（`oneflow.nn.functional.ctc_loss`函数不存在）
* 由于`ctc_loss`的`functor`函数比pytorch函数多了 `max_target_length` 这个参数，故不能直接在init文件中import cpp函数接口。我模仿`loss.py`中的写法，先定义了`max_target_length`，之后call c++函数
* 文档效果图
Screenshot 2023-02-27 at 10 39 17 AM

### CODE REVIEW

File: python/oneflow/test/modules/test_ctc_loss.py [hunk 1]
```diff
@@ -311,6 +311,29 @@ def _test_ctc_loss_with_diff_device_input(test_case, reduction):
     return out
 
 
+@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
+@autotest(n=3, auto_backward=True, check_graph=False)
+def _test_ctc_loss_functional_with_diff_device_input(test_case, reduction):
```
Comment (marigoold) [inline python/oneflow/test/modules/test_ctc_loss.py:316]: 这个 with_diff_device 是当时为了对齐 torch 而写的一个特例，这里如果测试 functional.ctc_loss 的话，不需要参考这个。
- 关于输入的 device，可以在 test/ 下面搜索 random_device 这个函数，看一下其他单测是怎么用的
- 同理，上面的 skipIf 也可以删掉了（这行的意思是，如果只测试 cpu 版本 oneflow 的话，就跳过这个测试，显然对于你这个单测的要求，这种行为是不合理的）
- tensor.tensor 这里可以用 random_tensor 来代替，也可以搜一下这个函数看一下其他单测怎么用的

**Comment by brandonliu2** [comment]
Screenshot 2023-02-27 at 10 39 17 AM
文档效果图

**Review by marigoold — APPROVED** [review]

File: python/oneflow/test/modules/test_ctc_loss.py [hunk 2]
```diff
@@ -323,6 +337,14 @@ def test_ctc_loss_with_diff_device_input(test_case):
         for arg in GenArgList(arg_dict):
             _test_ctc_loss_with_diff_device_input(test_case, *arg)
 
+    
+    def test_ctc_loss_functional(test_case):
+        arg_dict = OrderedDict()
+        arg_dict["reduction"] = ["mean", "none"]
+        for arg in GenArgList(arg_dict):
+            _test_ctc_loss_functional(test_case, *arg)
```
Comment (daquexian) [inline python/oneflow/test/modules/test_ctc_loss.py:345]: 这里调用 GenArgList 是多此一举的八股文写法，可以改掉，包括 test_ctc_loss_with_diff_device_input 里的：

```suggestion
        for reduction in ['mean', 'none']:
            _test_ctc_loss_functional(test_case, reduction)
```

即使这样改了之后，这里的遍历 `reduction in ['mean', 'none']` 也是没有必要的写法，reduction 的枚举可以在 autotest 里更优雅的完成

File: python/oneflow/test/modules/test_ctc_loss.py [hunk 3]
```diff
@@ -311,6 +311,20 @@ def _test_ctc_loss_with_diff_device_input(test_case, reduction):
...
 
+
+@autotest(n=3, auto_backward=True, check_graph=False)
+def _test_ctc_loss_functional(test_case, reduction):
+    device_random = random_device()
+    log_probs = random_tensor(ndim=3, dim0=5, dim1=2, dim2=3).to(device_random)
+    targets = random_tensor(ndim=2, dim0=2, dim1=3, low=1, high=3, dtype=int).to(device_random)
+    input_lengths = torch.tensor([5, 5], dtype=torch.int32)
+    target_lengths = torch.tensor([3, 3], dtype=torch.int32)
+    out = torch.nn.functional.ctc_loss(log_probs, targets, input_lengths, target_lengths, reduction=reduction)
```
Comment (daquexian) [inline python/oneflow/test/modules/test_ctc_loss.py:322]:
```suggestion
    
    out = torch.nn.functional.ctc_loss(log_probs, targets, input_lengths, target_lengths, reduction=oneof('mean', 'none', nothing())
```

oneof 是 autotest 提供的工具函数之一

File: python/oneflow/test/modules/test_ctc_loss.py [hunk 4]
```diff
@@ -317,12 +295,42 @@ def test_ctc_loss(test_case):
         for arg in gen_arg_list():
             compare_with_np(*arg)
 
+    @unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
+    # This test case can always success out of ci container, but will get error in ci container for unknown reason: error:
+    # 'oneflow.ctc_loss' op attribute 'blank' failed to satisfy constraint: 32-bit signed integer attribute
+    # loc("-":0:0): error: Failed to run round-trip passes
+    @autotest(n=5, auto_backward=True, check_graph=False)
```
Comment (daquexian) [inline python/oneflow/test/modules/test_ctc_loss.py:302]: 这里 auto_backward=True 不需要显式设置，auto_backward 默认就是 True，可以把两个测试一并改掉

**Review by daquexian — APPROVED** [review]
